### DISCUSSION TITLE
Telegraf -> Kafka -> Vector -> Prometheus

### DISCUSSION THREAD

**Discussion by sushovan23** [vectordotdev/vector#12773] [Q&A] [answered]
I'm trying to achieve the flow from telegraf -> kafka -> vector -> prometheus
In telegraf we have configured kafka output plugin and sending data in json format.

```
{
    "fields": {
        "field_1": 30,
        "field_2": 4,
        "field_N": 59,
        "n_images": 660
    },
    "name": "docker",
    "tags": {
        "host": "raynor"
    },
    "timestamp": 1458229140
}
```

Now in vector side. We have configured kafka source and able to read it in vector as verified by using `console` sink.
However my goal is to remote_write the kafka logs (which are essentially metrics) into prometheus using `remote_write` sink.

I read through documentation and there isn't a parser available for it. Is there anyone who has tried this scenario?

**Comment by jszwedko** [comment, marked as answer]
Hi @sushovan23 !

Can you provide examples of the metrics you want to write out in OpenMetrics format? For example:

```
field_1{host=raynor} 30 1458229140
```

? Or something else?

**Reply by sushovan23** [reply]
@jszwedko Thanks for the response!
My goal is to ultimately write it to prometheus using the sink of `prometheus_remote_write`

For example the json coming from kafka source would be like:
`{"fields":{"bytes_recv":284499084488,"bytes_sent":2919527673,"drop_in":0,"drop_out":0,"err_in":0,"err_out":0,"packets_recv":4513450331,"packets_sent":14277711},"name":"net","tags":{"host":"testvm.openstack","interface":"eth0"},"timestamp":1652905620}`

would be converted to prometheus metrics format:
```
net_bytes_recv{"host":"testvm.openstack","interface":"eth0"} 284499084488
net_bytes_sent{"host":"testvm.openstack","interface":"eth0"} 2919527673
```

so from my initial template in the question it would be
```
name_field1{"host": "raynor"} 30
...
name_fieldN{"host": "raynor"} nth_val
```

I'm new to prometheus_remote_write. so not sure how the timestamp will be added.

**Reply by jszwedko** [reply]
Thanks for the examples! I _think_ Vector doesn't currently have native support for translating that input into those metrics, but you can utilize the `lua` transform. It would look something like, in pseudocode:

```
[transforms.my_transform_id]
type = "lua"
inputs = [ "my-source-or-transform-id" ]
version = "2"
hooks.process = """
function(event, emit)
  for field, value in event.log.fields do
    emit {
      metric = {
        name = field,
        kind = "incremental",
        timestamp = // parse event.timestamp into a Lua timestamp,
        counter = {
          value = value
        }
        tags = event.log.tags,
      }
    }
  end
end
"""
```

This is rough Lua pseudocode but hopefully gives you something to go off of. This guide might also be useful: https://vector.dev/guides/advanced/custom-aggregations-with-lua/

**Reply by sushovan23** [reply]
where can I find more info about error:
`Configuration error. error=Transform "my_transform_id": Cannot evaluate Lua code defining "hooks.process": syntax error: [string "src/transforms/lua/v2/mod.rs:130:9"]:1: <name> expected near '('`

**Reply by jszwedko** [reply]
Hi @sushovan23 . That indicates you have a syntax error in your Lua script. I'd recommend trying to develop your script outside of Vector first (and just call it with `lua my-script.lua` and then plug it into Vector.

**Reply by sushovan23** [reply]
@jszwedko thanks for the pointers.

I have fixed the compiler errors now.
here is the function, it was missing name and ,

```
[transforms.my_transform_id]
type = "lua"
inputs = [ "kafka_input" ]
version = "2"
hooks.process = """
  function test(event, emit)
    for field, value in event.log.fields do
      emit {
        metric = {
          name = field,
          kind = "incremental",
          timestamp = event.timestamp,
          counter = {
            value = value
          },
          tags = event.log.tags
        }
      }
    end
  end
"""
```

Now I seem to be getting configuration error:
```
Configuration error. error=Transform "my_transform_id": Cannot evaluate Lua code defining "hooks.process": error converting Lua nil to function
```

**Reply by jszwedko** [reply]
Ah, it'll need to be an anonymous function when used there like:

```
[transforms.my_transform_id]
type = "lua"
inputs = [ "kafka_input" ]
version = "2"
hooks.process = """
  function(event, emit)
    for field, value in event.log.fields do
      emit {
        metric = {
          name = field,
          kind = "incremental",
          timestamp = event.timestamp,
          counter = {
            value = value
          },
          tags = event.log.tags
        }
      }
    end
  end
"""
```

(notice I deleted "test")

Alternatively, you can provide the Lua source and hooks separately like:

```
[transforms.my_transform_id]
type = "lua"
inputs = [ "kafka_input" ]
version = "2"
hooks.process = "test"
source = """
  function test(event, emit)
    for field, value in event.log.fields do
      emit {
        metric = {
          name = field,
          kind = "incremental",
          timestamp = event.timestamp,
          counter = {
            value = value
          },
          tags = event.log.tags
        }
      }
    end
  end
"""
```

Here you can see the source has the named function, `test`, which we then give to `hooks.process`.

**Reply by sushovan23** [reply]
I used anonymous function way now. No more Configuration Errors. Ran straight into `RuntimeErrorHooksProcess` 
`error=RuntimeErrorHooksProcess { source: RuntimeError("[string \"src/transforms/lua/v2/mod.rs:130:9\"]:2: for iterator 'for iterator' is not callable (a nil value)\nstack traceback:\n\t[C]: in for iterator 'for iterator'\n\t`
Tried changing `for field, value in event.log.fields do` to `for field, value in pairs(event.log.fields) do` 
also `for field, value in next, event.log.fields do`

error:
`vector::internal_events::lua: Error in lua script; discarding event. error=RuntimeErrorHooksProcess { source: RuntimeError("[string \"src/transforms/lua/v2/mod.rs:130:9\"]:2: bad argument #1 to 'for iterator' (table expected, got nil)\nstack traceback:\n\t[C]: in ?\n\t[C]: in function 'next'\n\t[string \"src/transforms/lua/v2/mod.rs:130:9\"]:2: in function <[string \"src/transforms/lua/v2/mod.rs:130:9\"]:1>") } error_type="script_failed" error_code="runtime_error_hook_process" stage="processing" internal_log_rate_secs=30`

**Reply by jszwedko** [reply]
I think you want to do `timestamp = event.log.timestamp` but also that you'll need to convert the unix timestamp into an actual Lua datetime table. I think you can use `os.date` for this. Something like `timestamp = os.date("*t", event.log.timestamp)` might work?

**Reply by sushovan23** [reply]
@jszwedko I feel the error however is unrelated to timestamp and more to do with how we are unpacking the json/table in lua
```
Error in lua script; discarding event. error=RuntimeErrorHooksProcess { source: RuntimeError("[string \"src/transforms/lua/v2/mod.rs:130:9\"]:2: bad argument #1 to 'for iterator' (table expected, got nil)
```

**Reply by jszwedko** [reply]
Oh I see, you are right, that does appear to be an error about the loop. I think the timestamp issue I mentioned above will _also_ be an issue, but that one seems to be related to the loop. I think you may need to do `for field, values in pairs(event.log.fields)` as shown on https://riptutorial.com/lua/example/2258/iterating-tables ? Again I'd recommend iterating on the Lua script in isolation first, calling your test function with input like:

```
test( { "log": {
    "fields": {
        "field_1": 30,
        "field_2": 4,
        "field_N": 59,
        "n_images": 660
    },
    "name": "docker",
    "tags": {
        "host": "raynor"
    },
    "timestamp": 1458229140
}})
```

That should make it easier to track down errors and resolve them.

**Reply by sushovan23** [reply]
@jszwedko I had tried that earlier it got lost in other things. but here is what I had searched and tried:

Tried changing `for field, value in event.log.fields do` to` for field, value in pairs(event.log.fields) do`
also for `field, value in next, event.log.fields do`

Yes I am trying to run it manually also. but was trying to see what input to pass since lua doesn't understand `json` unless im mistaken.

Found another thing. The kafka source which reads the metrics actually are looking different when I try to print the log using:
```
hooks.process = """
  function(event, emit)
    for log, log_value in pairs(event) do
      for key, value in pairs(log_value) do
         print(key, value)
      end
    end
  end
"""
```
this is what was printed from above:
```
headers	table: 0x109852d80
source_type	kafka
topic	mt_topic
[sources.kafka_input]
message_key
[sources.kafka_input]
timestamp	table: 0x109852d40
partition	4
offset	192657
message	{"fields":{"pid_count":1,"result_code":0,"running":1},"name":"procstat_lookup","tags":{"exe":"syslog","host":"myhost.openstack","pid_finder":"pgrep","result":"success"},"timestamp":1652917185}
```
so `event.log.fields` will not work as now message is also there. I tried with do `pairs(event.log.message.fields)` it also comes as nil.

I'm starting to believe that I would need to do nested for loops here to to messages and then do pairs for it.

**Reply by sushovan23** [reply]
@jszwedko kafka input source is making the message as a string instead of table.. so I am not able to unpack it using for loop in lua. 
any suggestions on how to extract the metrics?

**Reply by sushovan23** [reply]
Made a bunch of progress:

checked other post little similar to this and I added a `remap` before `lua` transform:

```
[transforms.my_test_parse]
  type = "remap"
  inputs = ["kafka_input"]
  source = '''
  .metric = parse_json!(.message)
  .metric.timestamp = to_timestamp!(.metric.timestamp)
  '''
```
now my next transform looks like this:

```
[transforms.my_transform_id]
type = "lua"
inputs = [ "my_test_parse" ]
version = "2"
hooks.process = """
  function(event, emit)
    for field, value in pairs(event.log.metric.fields) do
      emit {
        metric = {
          name = event.log.metric.name .. "_" .. field,
          kind = "incremental",
          timestamp = os.date("*t", XXXX),     -> problem here as now my timestamp is also a tableformat
          counter = {
            value = value
          },
          tags = event.log.metric.tags
        }
      }
    end
  end
"""
```
timestamp is a table which contains following:
```
year	2022
wday	5
yday	139
sec	15
isdst	false
min	59
month	5
day	19
hour	3
nanosec	0
```
I got no idea how to convert this to timestamp.

**Reply by sushovan23** [reply]
@jszwedko Thanks for all the help. `timestamp = os.time(event.log.metric.timestamp)`, was the function which takes table as input. 

Getting following error which doesn't really say what is wrong:
```
2022-05-19T04:23:05.794640Z ERROR transform{component_kind="transform" component_id=my_transform_id component_type=lua component_name=my_transform_id}: vector::internal_events::lua: Error in lua script; discarding event. error=RuntimeErrorHooksProcess { source: CallbackError { traceback: "stack traceback:\n\t[C]: in local 'emit'\n\t[string \"src/transforms/lua/v2/mod.rs:130:9\"]:3: in function <[string \"src/transforms/lua/v2/mod.rs:130:9\"]:1>", cause: FromLuaConversionError { from: "integer", to: "table", message: None } } } error_type="script_failed" error_code="runtime_error_hook_process" stage="processing" internal_log_rate_secs=30
2022-05-19T04:23:05.794876Z ERROR transform{component_kind="transform" component_id=my_transform_id component_type=lua component_name=my_transform_id}: vector::internal_events::lua: Internal log [Error in lua script; discarding event.] is being rate limited.
```


when I try with following:
`timestamp = event.log.metric.timestamp`
then get slightly different error
```
2022-05-19T04:39:45.961848Z ERROR transform{component_kind="transform" component_id=my_transform_id component_type=lua component_name=my_transform_id}: vector::internal_events::lua: Error in lua script; discarding event. error=RuntimeErrorHooksProcess { source: CallbackError { traceback: "stack traceback:\n\t[C]: in local 'emit'\n\t[string \"src/transforms/lua/v2/mod.rs:130:9\"]:3: in function <[string \"src/transforms/lua/v2/mod.rs:130:9\"]:1>", cause: FromLuaConversionError { from: "string", to: "f64", message: Some("expected number or string coercible to number") } } } error_type="script_failed" error_code="runtime_error_hook_process" stage="processing" internal_log_rate_secs=30
2022-05-19T04:39:45.976858Z ERROR transform{component_kind="transform" component_id=my_transform_id component_type=lua component_name=my_transform_id}: vector::internal_events::lua: Internal log [Error in lua script; discarding event.] is being rate limited.
```

**Reply by sushovan23** [reply]
@jszwedko Could you let me know where I can find about the formatting for timestamp in the emit:
```
      emit {
        metric = {
          name = event.log.metric.name .. "_" .. field,
          kind = "incremental",
          timestamp = os.date("*t", XXXX),     -> problem here as now my timestamp is also a tableformat
          counter = {
            value = value
          },
          tags = event.log.metric.tags
        }
      }
```

I have tried providing the timestamp in string format, in table format but nothing works

**Reply by sushovan23** [reply]
@jszwedko finally got it working!!! 

 `lua` transform I did this - `timestamp = event.log.metric.timestamp,`

seems like timestamp is taking data in table format

and I can see data in prometheus. 

Is there a way to make data expire after 60s or 120s. right now prometheus I can get data in graph but current time doesnt written since its not written yet. 

Guess I'm looking something similar to `expiration_interval` in telegraf output plugin. something similar for remote_write too are you aware
